### ISSUE TITLE
watson block seems to refresh only intermittently

### ISSUE THREAD

**Issue by quiescer** [greshake/i3status-rust#1305]
When I start tasks in watson I've found more often that not the watson block doesn't display them automatically. I can force the watson block to update by restarting i3 in place, which works, but I'd rather avoid having to do that.

When I run `i3status-rs ~/.config/i3status-rust/config-default.toml`in a terminal, I can see the Watson block is displayed and with the correct content.

**Comment by quiescer** [comment]
My config is below, although I've removed my API key and changed the location for the `openweathermap` block.

```
icons = "awesome5"
theme = "gruvbox-dark"
[[block]]
block = "watson"
show_time = true

[[block]]
block = "net"
format = "{speed_down;K*b} {speed_up;K*b} {graph_down}"
interval = 5

[[block]]
ap_format = "{strength}"
block = "networkmanager"
interface_name_exclude = ["docker\\d+"]
on_click = "kitty -e nmtui"
primary_only = true

[[block]]
alert = 10
alias = "/"
block = "disk_space"
format = "{icon} {used}/{total}"
info_type = "available"
interval = 60
path = "/"
unit = "GB"
warning = 20

[[block]]
block = "memory"
critical_mem = 95
critical_swap = 95
display_type = "memory"
format_mem = "{mem_used;G}/{mem_total;G}"
format_swap = "{swap_used;G}/{swap_total;G}"
warning_mem = 80
warning_swap = 80

[[block]]
block = "cpu"
format = "{barchart}"
interval = 1

[[block]]
block = "load"
format = "{1m}"
interval = 1

[[block]]
block = "temperature"
chip = "coretemp-*"
interval = 10

[[block]]
block = "sound"
driver = "pulseaudio"

[[block]]
block = "weather"
format = "{weather} {temp}C {humidity}%"

[block.service]
api_key = "XXXXXXXXXXXXXXXXXXXXXXXXXX"
name = "openweathermap"
place = "London,UK"
units = "metric"

[[block]]
block = "time"
format = "%a %d %b %R"
interval = 60
```

**Comment by ammgws** [comment]
> When I run `i3status-rs ~/.config/i3status-rust/config-default.toml`in a terminal, I can see the Watson block is displayed and with the correct content.

So when left running in the terminal it never stops updating like when run via i3bar?

**Comment by quiescer** [comment]
Yes, that's correct here's the moment when i3status-rs running in a terminal registered the change to Watson.
```
[{"full_text":"","color":"#282828","separator":false,"separator_block_width":0,"markup":"pango"},{"full_text":" INC1234 stopped after 1 week ","color":"#ebdbb2","background":"#282828","name":"0","instance":"0","separator":false,"separator_block_width":0,"markup":"pango"},{"full_text":"","color":"#282828","background":"#282828","separator":false,"separator_block_width":0,"markup":"pango"},{"full_text":"    4.9Kb   3.8Kb ▁▂█▃▁▅▃▂▃▂ ","color":"#ebdbb2","background":"#282828","name":"1","instance":"0","separator":false,"separator_block_width":0,"markup":"pango"},
<snip>
[{"full_text":"","color":"#282828","separator":false,"separator_block_width":0,"markup":"pango"},{"full_text":"","color":"#ebdbb2","background":"#282828","name":"0","instance":"0","separator":false,"separator_block_width":0,"markup":"pango"},{"full_text":"","color":"#282828","background":"#282828","separator":false,"separator_block_width":0,"markup":"pango"},{"full_text":"    4.9Kb   3.8Kb ▁▂█▃▁▅▃▂▃▂ ","color":"#ebdbb2","background":"#282828","name":"1","instance":"0","separator":false,"separator_block_width":0,"markup":"pango"},
```

**Comment by ammgws** [comment]
Can you change your bar command so that it `tee`'s the output to a file, that way we can observe the actual JSON at the time the bar stops updating?

Does the whole bar stop updating or only Watson?

Also are you on i3 or sway?

**Comment by quiescer** [comment]
Okay, I ran the bar instance as well as another instance in a terminal with both outputting to files. I ran both for for around 7 minutes, and immediately I saw the clock in my bar tick over to the next minute, I either started a new watson task (if none active) or stopped the currently active task. Only one or two of these state transitions seem to have been picked up by either instance of `i3status-rs`. Please see the attached files.

[i3status-rs.bar.log](https://github.com/greshake/i3status-rust/files/7274412/i3status-rs.bar.log)
[i3status-rs.term.log](https://github.com/greshake/i3status-rust/files/7274413/i3status-rs.term.log)
r

**Comment by quiescer** [comment]
Hmm, I've just noticed that the time on my bar's clock is lagging by a bit less than one minute also.

**Comment by anilanar** [comment]
I think there's a bug with something else.

TL;DR it works as long as you never `watson stop`.

Whenever I do `watson stop`, the block disappears and never comes back even after starting another project. However, if I prevent `i3status-rs` from picking up the stopped state by starting another project immediately after stopping, as in `watson stop && watson start another`, everything works fine. 

When I `watson stop`, it says `{projectname} stopped after {duration}` for a moment, and then it gets stuck with an empty string as in `full_text: ""`.

**Comment by ashleyghooper** [comment]
Yes, I believe you are on to something, @anilanar ! I just ran `inotifywait` recursively on my Watson config directory with `inotifywait -m -r .` then in another terminal, I ran:

```
watson stop; watson start Task19
```

There does **not** appear to be any CREATE event for the state file, as can be seen below.

```
./ OPEN state
./ ACCESS state
./ CLOSE_NOWRITE,CLOSE state
./ OPEN frames
./ ACCESS frames
./ CLOSE_NOWRITE,CLOSE frames
./ DELETE state.bak
./ MOVED_FROM state
./ MOVED_TO state.bak
./ MOVED_TO state
./ DELETE frames.bak
./ MOVED_FROM frames
./ MOVED_TO frames.bak
./ MOVED_TO frames
./ OPEN config
./ ACCESS config
./ CLOSE_NOWRITE,CLOSE config
./ OPEN state
./ ACCESS state
./ CLOSE_NOWRITE,CLOSE state
./ DELETE state.bak
./ MOVED_FROM state
./ MOVED_TO state.bak
./ MOVED_TO state
```

With the `inotifywait` still running I then ran `touch afile` in another terminal, which generated a CREATE:

```
./ CREATE afile
./ OPEN afile
./ ATTRIB afile
./ CLOSE_WRITE,CLOSE afile
```

It seems like the block's logic may not be quite suitable for the way watson runs. Perhaps we should wait for either `CREATE` or `MOVED_TO` events for the state file. For the record, I'm running this version of watson:

```
$ watson --version
Watson, version 2.0.1
```

**Comment by ammgws** [comment]
As expected - though not sure this ever worked for anyone if there's no CREATE events...

Anyway it seems reasonable to add MOVED_TO to the watch thread. Are you up to give it a go?

**Comment by MaxVerevkin** [comment]
@ashleyghooper what kernel do you use? What filesystem do you use?

> though not sure this ever worked for anyone if there's no CREATE events...

It works for me right now. 5.16.3-arch1-1 with ext4.

**Comment by ashleyghooper** [comment]
I've made a very simple commit that seems to enable Watson to show a new task on `watson start taskname` after having previously run a `watson stop`, however, from the point in that of the `watson stop` onwards, the block only updates when the state file changes, and so doesn't refresh the elapsed time in the task.

[Here](https://github.com/ashleyghooper/i3status-rust/commit/8d8d317aadf16db4e7dd18ac53499e0615a36f42) is a first attempt at a fix, but it's not a satisfactory fix yet due to the issue that the block switches to only updating when the state file changes, which leads me to suspect it could be related to the logic in lines 150-180:

```
            WatsonState::Idle {} => {                                                                                                                                                  
                if let Some(prev_state @ WatsonState::Active { .. }) = &self.prev_state {
...
```

**Comment by ammgws** [comment]
@MaxVerevkin do you get CREATE for the state file or only the state.bak file? Or both?

**Comment by ashleyghooper** [comment]
@MaxVerevkin, I'm using Linux 5.13.11 with btrfs filesystem. So the block continues updating for you if you do a `watson stop`, wait a minute, and then start a new task, with `watson start newtask`?

Do you also see a `CREATE` event for the `state` file if you run `inotify -m -r ~/.config/watson`, then stop your task and start a new one?

**Comment by MaxVerevkin** [comment]
> So the block continues updating for you if you do a watson stop, wait a minute, and then start a new task, with watson start newtask?

Yes.

> Do you also see a CREATE event for the state file if you run inotify -m -r ~/.config/watson, then stop your task and start a new one?

Yes.

> do you get CREATE for the state file or only the state.bak file? Or both?

I get CREATE only for `state` and `frames`.

**Comment by MaxVerevkin** [comment]
@ammgws do you use btrfs too?

**Comment by ammgws** [comment]
I don't/haven't actually run Watson. Just going off what I grokked from the source code. Though makes sense that it's due btrfs

**Comment by ashleyghooper** [comment]
@anilanar, do you use btrfs too?

**Comment by MaxVerevkin** [comment]
0b810cba985c9a8c83a2a705c949ef2949ad59fc should fix it.

> however, from the point in that of the watson stop onwards, the block only updates when the state file changes, and so doesn't refresh the elapsed time in the task.

I don't quite get that's wrong here.

**Comment by ashleyghooper** [comment]
Okay great, that's more concise than my initial fix, which retained the `match event.mask {` filter.

> I don't quite get what's wrong here.

Running `i3status-rs` in a terminal while running `watson` in another terminal, I find that the `i3status-rs` output ceases the first time I do a `watson stop`. From that time onwards, it will only update again when the `state` file changes, e.g. the below output is for a simple config which contains only a single watson block:

```{"version": 1, "click_events": true}
[
[{"full_text":" Task24 started now ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 1 minute ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 1 minute ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 1 minute ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 1 minute ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 1 minute ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 1 minute ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 2 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 2 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 2 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 2 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 2 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 2 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 3 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 3 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 3 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 3 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 3 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 3 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 4 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 4 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 4 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 4 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 4 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 4 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 5 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 5 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 5 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 5 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 5 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 5 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 started 6 minutes ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task24 stopped after 6 minutes ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task25 started now ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task25 started now ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task25 started now ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task25 stopped after 16 seconds ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":"","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task26 ","name":"0","instance":"0","markup":"pango"}],
```

The time between the 2nd to last line and the last line there was more than 30 minutes. `watson stop` causes the block - or perhaps the whole bar - to stop updating on a regular interval.

**Comment by ashleyghooper** [comment]
See also @anilanar's [comment](https://github.com/greshake/i3status-rust/issues/1305#issuecomment-1014504742) above

**Comment by MaxVerevkin** [comment]
Ah, I see. Does a8d3fd315b69116ef0be8ed78df83928f6a6e99d fix this?

**Comment by ashleyghooper** [comment]
Thanks @MaxVerevkin , unfortunately I've not been able to test it because I'm having trouble compiling `swayipc-types` all of a sudden, getting errors like the below. I am on NixOS, which complicates things slightly - I may try an alternative approach tomorrow.

```
error[E0277]: `[u8; 6]` is not an iterator
  --> /home/ash/.cargo/registry/src/github.com-1ecc6299db9ec823/swayipc-types-1.0.0/src/command.rs:24:33
   |
24 |         IntoIterator::into_iter(crate::MAGIC)
   |                                 ^^^^^^^^^^^^
   |                                 |
   |                                 expected an implementor of trait `IntoIterator`
   |                                 help: consider borrowing here: `&crate::MAGIC`
   |
   = note: the trait bound `[u8; 6]: IntoIterator` is not satisfied
   = note: required because of the requirements on the impl of `IntoIterator` for `[u8; 6]`
   = note: required by `into_iter`

error[E0599]: the method `chain` exists for array `[u8; 6]`, but its trait bounds were not satisfied
  --> /home/ash/.cargo/registry/src/github.com-1ecc6299db9ec823/swayipc-types-1.0.0/src/command.rs:25:14
   |
25 |             .chain(IntoIterator::into_iter(0_u32.to_ne_bytes()))
   |              ^^^^^ method cannot be called on `[u8; 6]` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `[u8; 6]: Iterator`
           which is required by `&mut [u8; 6]: Iterator`
           `[u8]: Iterator`
           which is required by `&mut [u8]: Iterator`
   = help: items from traits can only be used if the trait is in scope
   = note: the following trait is implemented but not in scope; perhaps add a `use` for it:
           `use std::io::Read;`
```

**Comment by ashleyghooper** [comment]
Okay, I suspect the problem I'm having compiling `swayipc-types` is a result of this change in `Cargo.toml`, possibly because of the version of my NixOS rust toolchain:

```
-swayipc = "3.0"
+swayipc = "2.7"
```

Anyway, I checked out a commit from around the time I lodged this issue and then did patched the changes to `src/blocks/watson.rs` from `master` branch in, but the time still stops updating after a `watson stop`, even when a subsequent `watson start taskname` is issued:

```
{"version": 1, "click_events": true}
[
[{"full_text":" Task26 started 1 hour ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task26 started 1 hour ago ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task26 stopped after 1 hour ","name":"0","instance":"0","markup":"pango"}],
[{"full_text":"","name":"0","instance":"0","markup":"pango"}],
[{"full_text":" Task27 started now ","name":"0","instance":"0","markup":"pango"}],
```

**Comment by MaxVerevkin** [comment]
Can you post a link to your version of `watson.rs`?

**Comment by ashleyghooper** [comment]
Here's what I did:
```
$ git checkout f146db87d68e357c05cf6ed6a91f2531ab4c606d
$ git checkout -p master src/blocks/watson.rs
# applied all hunks
```

**Comment by MaxVerevkin** [comment]
> I know, I'm naively counting on there not being changes elsewhere in the codebase that would affect this functionality :)

Well, there were some fixes since then (run `git diff master src/main.rs`).

**Comment by MaxVerevkin** [comment]
@ashleyghooper what's your rustc version?

**Comment by ashleyghooper** [comment]
> Well, there were some fixes since then (run `git diff master src/main.rs`).

And it seems like those fixes made a difference. I checked out the most recent commit before the change of `swayipc` version, which was `44b7cfca855f12efb4e8585bd990768ab289f4d4`, then I applied your update to `watson.rs` on top of it and tried again, and now it works!

Thanks for your help! Now I'll just need to create an override to get the current `master` branch version available as a Nix package. :)

For the record, my `rustc` version:
```
$ rustc --version
rustc 1.52.1
```
